use-cases/virgo/trainer.py

```diff
@@ -373,7 +367,278 @@ def train(self):
 
             # Report training metrics of last epoch to Ray
             train.report({"loss": np.mean(val_loss), "train_loss": np.mean(epoch_loss)})
+
+        return loss_plot, val_loss_plot, acc_plot, val_acc_plot
+
+
+class RayNoiseGeneratorTrainer(RayTorchTrainer):
+    def __init__(
+        self,
+        config: Dict,
+        strategy: Optional[Literal["ddp", "deepspeed"]] = "ddp",
+        name: Optional[str] = None,
+        logger: Optional[Logger] = None,
+    ) -> None:
+        super().__init__(config=config, strategy=strategy, name=name, logger=logger)
+
+    def create_model_loss_optimizer(self) -> None:
+        # Select generator
+        generator = self.training_config["generator"]
+        scaling = 0.02
+        if generator == "simple":
+            self.model = Decoder(3, norm=False)
+        elif generator == "deep":
+            self.model = Decoder_2d_deep(3)
+        elif generator == "resnet":
+            self.model = GeneratorResNet(3, 12, 1)
+            scaling = 0.01
+        elif generator == "unet":
+            self.model = UNet(input_channels=3, output_channels=1, norm=False)
+        else:
+            raise ValueError("Unrecognized generator type! Got", generator)
+
+        init_weights(self.model, "normal", scaling=scaling)
+
+        # Select loss
+        loss = self.training_config["loss"]
+        if loss == "l1":
+            self.loss = nn.L1Loss()
+        elif loss == "l2":
+            self.loss = nn.MSELoss()
+        else:
+            raise ValueError("Unrecognized loss type! Got", loss)
+
+        # Optimizer
+        self.optimizer = torch.optim.Adam(
+            self.model.parameters(), lr=self.training_config["learning_rate"]
+        )
+
+        # First, define strategy-wise optional configurations
+        if isinstance(self.strategy, RayDeepSpeedStrategy):
+            # Batch size definition is not optional for DeepSpeedStrategy!
+            distribute_kwargs = dict(
+                config_params=dict(
+                    train_micro_batch_size_per_gpu=self.training_config["batch_size"]
+                )
+            )
+        else:
+            distribute_kwargs = {}
+
+        # Distributed model, optimizer, and scheduler
+        self.model, self.optimizer, _ = self.strategy.distributed(
+            self.model, self.optimizer, **distribute_kwargs
+        )
+
+    def create_dataloaders(
```

Comment:
could this method be moved in the parent RayTorchTrainer?

Reply:
This is the same as in the other Virgo Trainer, because we use a custom collate function this needs to be redefined here